### PR TITLE
2036 cached test db templates

### LINKED ISSUE #2036: Cache a template test database to improve test run times
To avoid creating, migrating and populating a test database in every test, cache a template db and copy it for every test.

How it can work (description mainly for sqlite but could easily work for postgres as well):
- Have a test_cache directory somewhere in the repo
- Have a build.rs script that runs whenever a file in the migration dirs changes
- build.rs just deletes the test_cache directory
- When running a test, the test setup function checks if the cache dir exists. If it doesn't exist it build the template database
   - For the postgres case it can just recreates a template database and create an empty test_cache directory to mark that the postgres template database it is up to date...
- Copy template database to the test folder and run tests

### PR DESCRIPTION
**Pull request by clemens-msupply**
Fixes #2036 

# 👩🏻‍💻 What does this PR do? 
- Caches sqlite DB for testing (DBs are cached by: migration version and if all mock data is used)
- Invalidates cache when migrations or mock data changes (using build.rs)
- Fixes sqlite memory tests
- Updates postgres tests

## Overview of the cache logic:
- When migration or mock filed have been changed, a marker file is put in all existing test_output directory indicating that the cache needs to be invalidated
- When creating a test DB: If marker file exists delete existing template DBs and the marker file
- If no template DB exists create a new one and use it for the test

# 🧪 How has/should this change been tested? 
- Run sqlite tests (should already be faster than before)
- Second run should be much faster since all DB are cached now
- touch a migration file or a mock file, the cache should be invalidated and template file deleted before running next tests (this can be seen by, for example, look at the repository test_output directory while the test is running)

# Future work:
- Do similar work for Postgres
- Make more tests use full mock data to make more use of the cache
- Remove memory feature(?) since it would be the slowed option...

### CODE REVIEW

File: server/repository/src/test_db/sqlite.rs [hunk 1]
```diff
@@ -1,59 +1,157 @@
...
+                version.as_ref().unwrap_or(&Version::from_package_json())
+            )
+        } else {
+            format!(
+                "___template_{}",
+                version.as_ref().unwrap_or(&Version::from_package_json())
+            )
+        };
+
+        let guard = TEMPLATE_LOCK.lock().unwrap();
```
Comment (clemens-msupply) [inline server/repository/src/test_db/sqlite.rs:63]: Tests are running concurrently  and should not do fs operations at the same time.

File: server/service/src/test_helpers.rs [hunk 2]
```diff
@@ -17,6 +17,7 @@ pub(crate) struct ServiceTestContext {
     #[allow(dead_code)]
     pub(crate) connection: StorageConnection,
     pub(crate) service_provider: Arc<ServiceProvider>,
+    #[allow(dead_code)]
```
Comment (andreievg) [inline server/service/src/test_helpers.rs:20]: #[test] might be better

File: server/service/src/requisition_line/response_line_stats/mod.rs [hunk 3]
```diff
@@ -223,7 +223,7 @@ mod test {
     #[actix_rt::test]
     async fn get_response_requisition_line_stats_success() {
         let (_, _, connection_manager, _) = setup_all_with_data(
-            "get_requisition_line_chart_consumption",
+            "get_response_requisition_line_stats_success",
```
Comment (andreievg) [inline server/service/src/requisition_line/response_line_stats/mod.rs:226]: 🙏

File: server/repository/build.rs [hunk 4]
```diff
@@ -0,0 +1,18 @@
+use glob::glob;
+use std::fs;
+use std::path::Path;
+
+fn main() {
+    // when migrations are changing mark the template DBs to be recreated
+    for entry in glob("../**/test_output").expect("Failed to read glob pattern") {
+        match entry {
+            Ok(path) => {
+                fs::File::create(Path::new(&path).join("___template_needs_update.marker")).unwrap();
```
Comment (andreievg) [inline server/repository/build.rs:10]: We could just empty out cache here, might simplify it a little bit. 
Btw can share TEST_OUTPUT_DIR, TEMPLATE_MARKER_FILE constants with [the path attribute](https://doc.rust-lang.org/reference/items/modules.html#the-path-attribute).

File: server/repository/src/test_db/sqlite.rs [hunk 5]
```diff
@@ -1,59 +1,157 @@
...
 }
 
 pub async fn setup(db_settings: &DatabaseSettings) -> StorageConnectionManager {
-    setup_with_version(db_settings, None).await
+    setup_with_version(db_settings, None, MockDataInserts::none())
+        .await
+        .0
 }
 
+static TEMPLATE_LOCK: Mutex<()> = Mutex::new(());
```
Comment (andreievg) [inline server/repository/src/test_db/sqlite.rs:33]: that's pretty cool

File: server/repository/src/test_db/sqlite.rs [hunk 6]
```diff
@@ -1,59 +1,157 @@
...
+            format!(
+                "___template_{}",
+                version.as_ref().unwrap_or(&Version::from_package_json())
+            )
+        };
+
+        let guard = TEMPLATE_LOCK.lock().unwrap();
+        // if marker exists, DB needs to be recreated -> delete all template files
+        let marker_path =
+            Path::new(&format!("{}/{}", TEST_OUTPUT_DIR, TEMPLATE_MARKER_FILE)).to_path_buf();
```
[the remark this answers is not in the document]
Comment (andreievg) [reply server/repository/src/test_db/sqlite.rs:66]: This would also allow us to use `std::sync::Once` or [ctor](https://docs.rs/ctor/latest/ctor/), and create both all mock and just db migration file (if they don't exist), it should simplify this bit of logic (or at least compartmentalise it)

**Review by andreievg — APPROVED** [review]
Looks very good, thanks, and super quick.

I have some suggestions, I think they can be captured in another issue (inlucding future improvement suggestions that you've added to PR description). A

Thanks for test cases, followed them and all pass.

I was thinking of a use case where this may break, can't really think of any (potentially when editing mock data while tests are starting, but that's super edge case and restarting  tests should just fix it)

File: server/repository/src/test_db/sqlite.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (clemens-msupply) [reply server/repository/src/test_db/sqlite.rs:66]: Yes that what I was thinking as well but the different test run in different processes. So we would need a file lock to synchronise the cache creation between the processes (e.g. there is a proc-lock crate). But then I thought its too complicated for the first version. But I think I could actually use std::sync:Once instead of the TEMPLATE_LOCK.

File: server/repository/build.rs (hunk 4 above)
Comment (clemens-msupply) [reply server/repository/build.rs:10]: Thanks did this, i.e. I moved them into a constants file

Comment (clemens-msupply) [reply server/repository/build.rs:10]: I kept the code in the build.rs minimal because I found it slightly hard to debug and to test the build.rs file (e.g. because of the re-run conditions)... Furthermore, I was also thinking about reusing the marker for Postgres (e.g. Postgres rebuilds template DB when the `test_output` dir doesn't exist OR if there is a marker file in the `test_output` dir)

File: server/repository/src/test_db/sqlite.rs [hunk 7]
```diff
@@ -1,59 +1,157 @@
...
+        if marker_path.exists() {
+            // remove all DB templates
+            for entry in fs::read_dir(TEST_OUTPUT_DIR).unwrap() {
+                let entry = entry.unwrap();
+                if entry.file_name().to_string_lossy() == TEMPLATE_MARKER_FILE {
```
[the remark this answers is not in the document]
Comment (clemens-msupply) [reply server/repository/src/test_db/sqlite.rs:71]: Yes good point, was about to change but couldn't come up with a better name and then thought it might be good to explicitly ensure we don't delete the file before cleaning up the templates...

**Review by jmbrunskill — APPROVED** [review]
Nice work Clemens, just took a quick look at this stage, but like the idea.
Always good to avoid duplicate effort like migrations & mock data!

File: server/service/src/test_helpers.rs (hunk 2 above)
Comment (clemens-msupply) [reply server/service/src/test_helpers.rs:20]: The module of the ServiceTestContext is already `#[cfg(test)]`. This was an issue for the `memory` tests because I disabled the processor tests for them and then this field became unused. (Tried to mark it with `not(memory)` but this had lots of other issues...)

**Comment by clemens-msupply** [comment]
Thanks for the review! looking forward to do the Postgres test on the next R&D day!
